blocks/bricks/bn.py

```diff
@@ -0,0 +1,345 @@
+    defaults (unit scale and zero offset), but you can explicitly pass one
+    or both initializers to override this.
+
+    This has the necessary properties to be inserted into a
+    :class:`blocks.bricks.conv.ConvolutionalSequence` as-is, in which case
+    the `input_dim` should be omitted at construction, to be inferred from
+    the layer below.
+
+    """
+    @lazy(allocation=['input_dim'])
+    def __init__(self, input_dim, broadcastable=None,
+                 save_memory=True, epsilon=1e-4, weights_init=None,
+                 biases_init=None, **kwargs):
+        self.input_dim = input_dim
+        self.broadcastable = broadcastable
+        self.save_memory = save_memory
+        self.epsilon = 1e-4
+        self.weights_init = (Constant(1) if weights_init is None
+                             else weights_init)
+        self.biases_init = (Constant(0) if biases_init is None
+                            else biases_init)
+        self._training_mode = False
+        super(BatchNormalization, self).__init__(**kwargs)
+
+    @application(inputs=['input_'], outputs=['output'])
+    def apply(self, input_, application_call):
+        if self._training_mode:
+            mean, stdev = self._compute_training_statistics(input_)
+        else:
+            mean, stdev = self._prepare_population_statistics()
+        # Useful for filtration of calls that were already made in
+        # training mode when doing graph transformations.
+        application_call.metadata['training_mode'] = self._training_mode
+        # Useful for retrieving a list of updates for population
+        # statistics. Ditch the broadcastable first axis, though, to
+        # make it the same dimensions as the population mean and stdev
+        # shared variables.
+        application_call.metadata['offset'] = mean[0]
+        application_call.metadata['divisor'] = stdev[0]
+        # Give these quantities roles in the graph.
+        _add_role_and_annotate(mean, BATCH_NORM_OFFSET,
+                               [self, application_call])
+        _add_role_and_annotate(stdev, BATCH_NORM_DIVISOR,
+                               [self, application_call])
+        W = _add_batch_axis(self.W, "W.dimshuffle('x'...)")
+        b = _add_batch_axis(self.b, "b.dimshuffle('x', ...)")
+        # Heavy lifting is done by the Theano utility function.
+        normalized = bn.batch_normalization(input_, W, b, mean, stdev,
+                                            mode=('low_mem' if self.save_memory
+                                                  else 'high_mem'))
+        return normalized
+
+    def __enter__(self):
+        self._training_mode = True
+
+    def __exit__(self, *exc_info):
+        self._training_mode = False
+
+    def _compute_training_statistics(self, input_):
+        axes = (0,) + tuple((i + 1) for i, b in
+                            enumerate(self.population_mean.broadcastable)
+                            if b)
```

Comment:
That actually won't work, as self.broadcastable can be `None` as per the current `_allocate`.